crates/bevy_render/src/mesh/mesh/mod.rs

```diff
@@ -314,6 +351,132 @@ impl Mesh {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct MeshVertexAttribute {
+    /// The friendly name of the vertex attribute
+    pub name: &'static str,
+
+    /// The _unique_ id of the vertex attribute. This will also determine sort ordering
+    /// when generating vertex buffers. Built-in / standard attributes will use "close to zero"
+    /// indices. When in doubt, use a random / very large usize to avoid conflicts.
+    pub id: MeshVertexAttributeId,
+
+    /// The format of the vertex attribute.
+    pub format: VertexFormat,
+}
+
+impl MeshVertexAttribute {
+    pub const fn new(name: &'static str, id: usize, format: VertexFormat) -> Self {
+        Self {
+            name,
+            id: MeshVertexAttributeId(id),
+            format,
+        }
+    }
+
+    pub const fn at_shader_location(&self, shader_location: u32) -> VertexAttributeDescriptor {
+        VertexAttributeDescriptor::new(shader_location, self.id, self.name)
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Ord, PartialOrd, Hash)]
+pub struct MeshVertexAttributeId(usize);
+
+impl From<MeshVertexAttribute> for MeshVertexAttributeId {
+    fn from(attribute: MeshVertexAttribute) -> Self {
+        attribute.id
+    }
+}
+
+pub type MeshVertexBufferLayout = Hashed<InnerMeshVertexBufferLayout>;
+
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub struct InnerMeshVertexBufferLayout {
+    attribute_ids: Vec<MeshVertexAttributeId>,
+    layout: VertexBufferLayout,
+}
+
+impl InnerMeshVertexBufferLayout {
+    #[inline]
+    pub fn contains(&self, attribute_id: impl Into<MeshVertexAttributeId>) -> bool {
+        self.attribute_ids.contains(&attribute_id.into())
+    }
+
+    // TODO: maybe these should be hidden to ensure specializers can only use this type to generate VertexBufferLayouts?
```

Comment:
Just a comment to remind of the existence of this todo

Reply:
Ah yeah. Removing the ability to extract any information directly from MeshVertexBufferLayout protects against the corner case of "using information that doesn't get encoded in the final VertexBufferLayout to specialize the pipeline, which could break the guarantees of the cache". For example, checking if a MeshVertexBufferLayout contains an attribute that isn't used in the shader and using that information to enable a "shader def". This would result in odd behavior because that pipeline would be returned for a (Key, MeshVertexBufferLayout) pair that _doesnt_ have that attribute, but results in an equivalent (Key, VertexBufferLayout) pair.

In practice, I don't think this will be an issue because if you are enabling behaviors based on vertex attribute existence, you will be using them in your shaders. And including these methods (especially `contains()`) make for really straightforward pipeline construction (`contains()` allowed us to simplify the MeshPipelineKey and GpuMesh by removing `has_tangents` flags).

We could "resolve" this by:

1. Removing VertexBufferLayout de-duping (not really an option ... this could introduce significant overhead in some scenes)
2. Removing the ability to read information from MeshVertexBufferLayout directly. This would require passing that information in via flags, or using "trial and error" to try a number of different layouts to see what is available. And even then, you could use the "trial and error" approach to leak information. So hiding methods just makes it _harder_ to leak information, not impossible.
3. Rephrase the specialization api to decouple the `(Key, MeshVertexBufferLayout) -> VertexBufferLayout` conversion from `(Key, VertexBufferLayout) -> RenderPipelineDescriptor`. This would be a very "opinionated" api that might not work for every scenario (ex: instancing) and makes everything just a bit more complicated.
4. Compare the entire pipeline with the "cached" pipeline to detect (and fail on) cases where "unused mesh vertex buffer layout" information is used specialize a pipeline. This could be expensive in practice because we would need to do a full comparison of every piece of information in the pipeline for each potentially duplicated pipeline.

Reply:
I chose to implement (4), but as a "debug assertion" to avoid incurring significant overhead in released games. We can revisit this behavior if it ends up holding back debug builds too much (which I doubt ... I'm guessing that deduplicating will only happen in non-standard workflows). I also verified that it logs an error in the appropriate case.